database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```
@@ -0,0 +1,170 @@
+    }
+
+    @Override
+    public void onChildAdded(DataSnapshot keySnapshot, String previousChildKey) {
+        super.setOnChangedListener(null);
+        super.onChildAdded(keySnapshot, previousChildKey);
+        super.setOnChangedListener(mListener);
+
+        DatabaseReference ref = mRef.child(keySnapshot.getKey());
+        mRefs.put(ref, ref.addValueEventListener(new DataRefSnapshot()));
+    }
+
+    @Override
+    public void onChildChanged(DataSnapshot snapshot, String previousChildKey) {
+        super.setOnChangedListener(null);
+        super.onChildChanged(snapshot, previousChildKey);
+        super.setOnChangedListener(mListener);
+    }
+
+    @Override
+    public void onChildRemoved(DataSnapshot keySnapshot) {
+        String key = keySnapshot.getKey();
+        int index = getIndexForKey(key);
+        mRef.child(key).removeEventListener(mRefs.remove(mRef.child(key)));
+
+        super.setOnChangedListener(null);
+        super.onChildRemoved(keySnapshot);
+        super.setOnChangedListener(mListener);
+
+        if (isMatch(index, key)) {
+            mDataSnapshots.remove(index);
+            notifyChangedListeners(OnChangedListener.EventType.REMOVED, index);
+        }
+    }
+
+    @Override
+    public void onChildMoved(DataSnapshot keySnapshot, String previousChildKey) {
+        String key = keySnapshot.getKey();
+        int oldIndex = getIndexForKey(key);
+
+        super.setOnChangedListener(null);
+        super.onChildMoved(keySnapshot, previousChildKey);
+        super.setOnChangedListener(mListener);
+
+        if (isMatch(oldIndex, key)) {
+            DataSnapshot snapshot = mDataSnapshots.remove(oldIndex);
+            int newIndex = getIndexForKey(key);
+            mDataSnapshots.add(newIndex, snapshot);
+            notifyChangedListeners(OnChangedListener.EventType.MOVED, newIndex, oldIndex);
+        }
+    }
+
+    @Override
+    public void onCancelled(DatabaseError error) {
+    }
```

Here I'm completely ignoring errors from the indexed key refs